### PR TITLE
Stopped loading Figwheel API by default #2669

### LINKED ISSUE #2669: Remove dependency on user.clj
### User Story

As a developer, I want to have a predictable environment.

### Description

`user.clj` is loaded automatically by clojure runtime. This loading can't be controlled.
This is an issue as we are using it to setup and load figwheel dependencies even when it's not required.
e.g. `lein cljsbuild once android` never terminates 

The modern way to manage complex `figwheel` workflow is to rely on [figwheel API](https://github.com/bhauman/lein-figwheel#scripting-figwheel) and run it via `lein run -m clojure.main --init`. This will also allow to have all figwheel related code in a single location (currently spread in `env/dev` and `user.clj`).

*Type*: Feature

### PR DESCRIPTION
**Pull request by ghost**
Fixes #2669

### Summary:
Loads Figwheel API on demand, but not automatically.

### Steps to test 0:
- in shell run 'lein figwheel-repl ios android'
- or 'lein figwheel-repl ios'
- or 'lein figwheel-repl android'
- or 'lein figwheel-repl'   (defaults to android)
- cljs-repl runs

### Steps to test 1:
- run Leiningen REPL: "lein repl'
- functions such as 'start-figwheel should not load.
- all other Leiningen tasks should be figwheel free.

status: ready

### CODE REVIEW

**Comment by poiga** [comment]
I have been investigating the problem and so far found that user.clj is the common way to do keep functions for development services. https://github.com/bhauman/lein-figwheel/blob/3350938413ba5e6ed53026a5a82ba6321314ee2f/examples/separate-server/dev/user.clj 
http://practicalli.github.io/clojurescript/reagent-projects/tic-tac-toe/figwheel-helper-functions.html
and 'user' is default namespace for many REPLs.
Although I have made loading of figwheel APIs optional.

File: env/dev/env/figwheel_api.clj [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    (fn [id]
+      (assoc
+        (let [build (get all-builds (get-id id))]
+          (if (test-id? id)
+            (get-test-build build)
+            build))
+        :id id))
+    ids))
+
+(let [env-build-ids (System/getenv "BUILD_IDS")
```
Comment (jeluard) [inline env/dev/env/figwheel_api.clj]: Can this be wrapped in a function so that we can control its lifecycle? Same for `hawk/watch!`

**Review by jeluard — CHANGES_REQUESTED** [review]
Nice job!
`user.clj` might be commonly used but that's a bad pattern IMHO. I'd rather have explicit configuration / function call than magic behavior.

Also we want to keep a simple command line oriented flow so everything should be ready after a `lein` call.

What about using `lein run -m clojure.main --init` to launch the figwheel logic? This complexity would then be hidden behind a `lein dev` alias?.

**Comment by poiga** [comment]
@jeluard 

> This complexity would then be hidden behind a lein dev alias.

I think it is now :) please check!

**Comment by jeluard** [comment]
I am curious: what are the advantages of using repl hooks vs the method detailed by [figwheel documentation](https://github.com/bhauman/lein-figwheel#scripting-figwheel) ? Are there any startup cost for instance?

**Comment by ghost** [comment]
Just your previous comment contained "dev alias" so I thought that will be good Idea.
I just changed the commit and now we have it as a script(not sure about it's right location) rather than REPL hook. But I left 2 REPL aliases, as I think they are optional to use, but may be convenient to have. Your opinion?

**Comment by jeluard** [comment]
With current solution, does `lein cljsbuild once android` terminates (i.e. figwheel is not started) ?

**Comment by jeluard** [comment]
@poiga Sorry about the confusion, by `dev alias` I meant using lein alias to encapsulate `lein run -m clojure.main env/dev/figwheel_api.clj'`

**Comment by poiga** [comment]
Yes, 'lein cljsbuild once android' does terminate and Figwheel is not started.

**Comment by ghost** [comment]
Yes, 'lein cljsbuild once android' does terminate and Figwheel is not started.
And I added 'lein dev-repl' alias to hide the script.

**Comment by ghost** [comment]
@jeluard Do you think it is OK now?

**Review by jeluard — APPROVED** [review]

**Comment by jeluard** [comment]
Looks like yes!

**Comment by jeluard** [comment]
@poiga Running `lein cljsbuild once` I can see the following: `Figwheel: Starting server at http://0.0.0.0:3449` (and build never stops).

Also I am not clear when I want to use `lein dev-repl` vs `lein dev-ios` ?

**Review by jeluard — CHANGES_REQUESTED** [review]
Figwheel still runs

**Comment by ghost** [comment]
do you really have the latest? :O

**Comment by jeluard** [comment]
Yes I just cloned your repo. Have you tried `lein cljsbuild once`, seen compilation but without figwheel? (you might need to clean target)

**Comment by poiga** [comment]
I did not notice that one.
Thanks for testing!
Fixed now.

**Comment by ghost** [comment]
@jeluard I did not notice that one.
Thanks for testing!
Fixed now.

File: project.clj [hunk 2]
```diff
@@ -30,14 +31,14 @@
                     :source-paths ["src" "env/dev"]
```
Comment (jeluard) [inline project.clj:31]: Is this still needed?

**Review by jeluard — APPROVED** [review]
Great! Let's wait after christmas to get some more feedback and other people testing it.

File: project.clj (hunk 2 above)
Comment (poiga) [reply project.clj]: well, everything still works, when I remove "src" part, but not "env/dev". I will try to investigate into this.

Comment (ghost) [reply project.clj:31]: looks like these are in place now.

**Comment by ghost** [comment]
Is here something I could do to help testing or something else?

**Comment by jeluard** [comment]
@poiga Looks good to me!

**Comment by dmitryn** [comment]
@poiga thanks for contribution! Do you think this line should be changed as well? https://github.com/status-im/status-react/blob/develop/scripts/run-osx#L67

**Comment by ghost** [comment]
@dmitryn Thanks for noticing! I also fixed few other locations for the same thing. I will fix "Building Status" page in Wiki just after merging as well.

**Comment by poiga** [comment]
@dmitryn Thanks for noticing! I also fixed few other locations for the same thing. I will fix "Building Status" page in Wiki just after merging as well.

**Review by dmitryn — APPROVED** [review]

**Comment by ghost** [comment]
@dmitryn @jeluard it would be cool to merge as well :)

**Comment by jeluard** [comment]
@poiga We currently have issues with our internal Jenkins. We'll merge as soon as it is operational again !

File: env/dev/env/config.cljs [hunk 3]
```diff
@@ -2,5 +2,5 @@
 
 (def figwheel-urls {
                     :ios "ws://localhost:3449/figwheel-ws"
-                    :android "ws://10.0.2.2:3449/figwheel-ws"
+                    :android "ws://localhost:3449/figwheel-ws"
```
Comment (jeluard) [inline env/dev/env/config.cljs]: This file should not be part of the commit.

**Comment by jeluard** [comment]
@poiga Jenkins build passed! Please address my last comment (about unneeded change) and we will merge.

**Comment by ghost** [comment]
@jeluard I removed the file. Thanks

**Review by flexsurfer — APPROVED** [review]

**Comment by flexsurfer** [comment]
thanks for contribution @poiga , is it possible to get rid of BUILD_IDS in this PR? and have something like `lein dev-repl-ios` or better `lein dev-repl ios`  and `lein dev-repl android`, thanks

**Comment by pablodip** [comment]
@poiga Would you mind to rename "dev-repl" to "figwheel-repl" as well, please? Thanks! :)

**Comment by ghost** [comment]
@flexsurfer I can, but @jeluard did not think this is good idea when I did it last time. Are you sere we need it?

**Comment by ghost** [comment]
So according to @pablodip and @flexsurfer comments - I am going to add lein aliases:
lein figwheel-repl ios
lein figwheel-repl android

Is that You think right?

**Comment by flexsurfer** [comment]
@poiga yes, we discussed this with him, are you sure we can't use argument? `lein figwheel-repl ios`

**Comment by jeluard** [comment]
@poiga Sorry for the confusion! Yes as @flexsurfer said if we can have `lein figwheel-repl` accept an argument (default to `android`) and drop the env variable support that would be awesome!

Thanks again!

**Comment by ghost** [comment]
@jeluard @flexsurfer @pablodip all done! thanks for testing ;)

**Comment by foopang** [comment]
It seems that hot reloading is broken after this. I need to do `(reset-autobuild)`, then close and open the app in emulator in order to see new changes applied.

File: project.clj [hunk 4]
```diff
@@ -30,14 +31,14 @@
...
                     :cljsbuild    {:builds
                                    {:ios
-                                    {:source-paths ["react-native/src" "src" "env/dev"]
+                                    {:source-paths ["react-native/src"]
```
Comment (foopang) [inline project.clj:34]: This makes hot reloading fail.

**Comment by oskarth** [comment]
@foopang Thanks for report! I was looking at this specific diff set before, and while I haven't validated yet I would agree with you. Would you mind opening an issue for this?

EDIT: Oh I see you just filed a PR, great! https://github.com/status-im/status-react/pull/2898

**Comment by oskarth** [comment]
cc @jeluard @flexsurfer @dmitryn for how we can catch this in future reviews
